pep-0584.rst

```diff
@@ -1,1043 +1,974 @@
+subclasses of dict.
 
-    all follow the same rule.  This PEP takes the position that this
-    behaviour is simple, obvious, and usually the behaviour we want,
-    and should be the default behaviour for dicts.
+Further, ChainMaps wrap their underlying dicts, so writes to the
+ChainMap will modify the original dict::
 
-2.  **Raise on conflicting keys.**  It isn't clear that this behaviour has
-    many use-cases or will be often useful, but it will likely be annoying
-    as any use of the dict plus operator would have to be guarded with a
-    ``try...except`` clause.
+      >>> d1 = {'spam': 1}
+      >>> d2 = {'eggs': 2}
+      >>> merged = ChainMap(d2, d1)
+      >>> merged['eggs'] = 999
+      >>> d2
+      {'eggs': 999}
 
-    But for those who want it, it would be easy to override the method in a
-    subclass to get the desired behaviour::
 
-        def __add__(self, other):
-            if self.keys() & other.keys():
-                raise KeyError('duplicate key')
-            return super().__add__(other)
+=========
+Rationale
+=========
 
-3.  **Add the values (as Counter does).**  Too specialised to be used as
-    the default behaviour.
+The new operators will have the same relationship to the
+``dict.update`` method as the list concatenate (``+``) and extend
+(``+=``) operators have to ``list.extend``.
 
-4.  **First seen wins: value-preserving semantics.**  It isn't clear that
-    this behaviour has many use-cases.  Probably too specialised to be used
-    as the default, best to leave this for subclasses, or simply reverse
-    the order of the arguments::
+Key conflicts will be resolved by keeping the rightmost value.  This
+matches the existing behavior of similar ``dict`` operations, where
+the last seen value always wins::
 
-        # d1 merged with d2, keeping existing values in d1
-        d2 + d1
+    {'a': 1, 'a': 2}
+    {**d, **e}
+    d.update(e)
+    d[k] = v
+    {k: v for x in (d, e) for (k, v) in x.items()}
 
-5.  **Concatenate values in a list**::
+All of the above follow the same rule.  This PEP takes the position
+that this behavior is simple, obvious, usually the behavior we want,
+and should be the default behavior for dicts.  This means that dict
+union is not commutative; in general ``d | e != e | d``.
 
-        {'a': 1} + {'a': 2} == {'a': [1, 2]}
+Similarly, the *iteration order* of the key-value pairs in the
+dictionary will follow the same semantics as the examples above, with
+each newly merged pair being appended (or moved) to the end of the
+current sequence.
 
-    This is likely to be too specialised to be the default.  It is not clear
-    what to do if the values are already lists::
 
-        {'a': [1, 2]} + {'a': [3, 4]}
+=============
+Specification
+=============
 
-    Should this give ``{'a': [1, 2, 3, 4]}`` or ``{'a': [[1, 2], [3, 4]]}``?
+Dict union will return a new dict containing the left operand merged
+with the right operand. If a key appears in both operands, the
```

Comment:
This doesn't specify constraints on the RHS type. The pseudo code below indicates it should be a dict. Let's be explicit about that here.